src/main/java/com/emc/pravega/perf/PravegaPerfTest.java

```diff
@@ -400,76 +602,70 @@ public void run() {
                 return null;
 
           };
-        }
-    }
+      }
+}
 
-    /**
-     * A Sensor reader class that reads the temperative data
-     */
-    private static class ReaderWorker implements Runnable {
-       @Setter
-        public static AtomicInteger totalEvents;
-        private EventStreamReader<String> reader;
-        String readerId;
+class ReaderWorker implements Callable<Void> {
+       public static AtomicInteger totalEvents;
+       private EventStreamReader<String> reader;
+       private final int secondsToRun;
+       private final long StartTime;
+       private final int timeout;
+       private final PerfStats stats;
+       String readerId;
 
 
-        public ReaderWorker(int readerId) {
+       ReaderWorker (int readerId, int secondsToRun, long start,
+                    ClientFactory factory, PerfStats stats,
+                    String readergrp, int timeout) {
             this.readerId = Integer.toString(readerId);
-            ClientFactory clientFactory = ClientFactory.withScope("Scope",
-                    URI.create(controllerUri));
-
-            reader = clientFactory.createReader(
-                    this.readerId, streamName, new JavaSerializer<String>(), ReaderConfig.builder().build());
+            this.secondsToRun = secondsToRun; 
+            this.StartTime=start;
+            this.stats = stats; 
+            this.timeout = timeout;
+
+            reader = factory.createReader(
+                    this.readerId, readergrp, new JavaSerializer<String>(), ReaderConfig.builder().build());
+       }
+
+       void cleanupEvents(PerfStats drainStats) throws Exception {
```

Comment:
Private. Also Narrow this exception type.

Reply:
i want to keep this generic type, are you see any problem with this?

Reply:
This is not generic. If you want to parameter your exception go ahead. This is forcing a single catch block on the caller where they don't know how to handle things in a differentiated way. If your only intention is to signal an unrecoverable problem, then you can throw RuntimeException and you don't need this on the signature at all. If you want to signal a specific handle-able problem then declare the type. 
See our guide:
https://github.com/pravega/pravega/wiki/Contributing#exceptions

Reply:
yes, i have taken care this now.. all the methods will throw the specific errors